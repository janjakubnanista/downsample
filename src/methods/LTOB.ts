import { DownsamplingFunction, DownsamplingFunctionConfig, NormalizedDataPoint } from '../types';
import { calculateTriangleArea, createLegacyDataPointConfig, createNormalize } from '../utils';

// Largest triangle three buckets data downsampling algorithm implementation
export const createLTOB = <P>(config: DownsamplingFunctionConfig<P>): DownsamplingFunction<P, [number]> => {
  const normalize = createNormalize(config.x, config.y);

  return (data: P[], desiredLength: number): P[] => {
    if (desiredLength < 0) {
      throw new Error(`Supplied negative desiredLength parameter to LTOB: ${desiredLength}`);
    }

    const { length } = data;
    if (length <= 1 || length <= desiredLength) {
      return data;
    }

    // Now we are sure that:
    //
    // - length is [2, Infinity)
    // - threshold is (length, Inifnity)
    const bucketSize: number = length / desiredLength;
    const normalizedData: NormalizedDataPoint[] = normalize(data);
    const sampledData: P[] = [data[0]];

    for (let bucket = 1; bucket < desiredLength - 1; bucket++) {
      const startIndex: number = Math.floor(bucket * bucketSize);
      const endIndex: number = Math.min(length - 1, (bucket + 1) * bucketSize);

      let maxArea = -1;
      let maxAreaIndex = -1;
      for (let j: number = startIndex; j < endIndex; j++) {
        const previousDataPoint: NormalizedDataPoint = normalizedData[j - 1];
        const dataPoint: NormalizedDataPoint = normalizedData[j];
        const nextDataPoint: NormalizedDataPoint = normalizedData[j + 1];

        const area = calculateTriangleArea(previousDataPoint, dataPoint, nextDataPoint);
        if (area > maxArea) {
          maxArea = area;
          maxAreaIndex = j;
        }
      }

      sampledData.push(data[maxAreaIndex]);
    }

    sampledData.push(data[length - 1]);

    return sampledData;
  };
};

export const LTOB = createLTOB(createLegacyDataPointConfig());
