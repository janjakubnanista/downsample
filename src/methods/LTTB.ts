import { DownsamplingFunction, DownsamplingFunctionConfig, NormalizedDataPoint } from '../types';
import {
  calculateAverageDataPoint,
  calculateTriangleArea,
  createLegacyDataPointConfig,
  createNormalize,
  splitIntoBuckets,
} from '../utils';

export function LTTBIndexesForBuckets(buckets: NormalizedDataPoint[][]): number[] {
  const bucketCount: number = buckets.length;
  const bucketDataPointIndexes: number[] = [0];

  let previousBucketsSize = 1;
  let lastSelectedDataPoint: NormalizedDataPoint = buckets[0][0];
  for (let index = 1; index < bucketCount - 1; index++) {
    const bucket: NormalizedDataPoint[] = buckets[index];
    const nextBucket: NormalizedDataPoint[] = buckets[index + 1];
    const averageDataPointFromNextBucket = calculateAverageDataPoint(...nextBucket);
    if (averageDataPointFromNextBucket === undefined) continue;

    let maxArea = -1;
    let maxAreaIndex = -1;
    for (let j = 0; j < bucket.length; j++) {
      const dataPoint: NormalizedDataPoint = bucket[j];
      const area = calculateTriangleArea(lastSelectedDataPoint, dataPoint, averageDataPointFromNextBucket);

      if (area > maxArea) {
        maxArea = area;
        maxAreaIndex = j;
      }
    }

    lastSelectedDataPoint = bucket[maxAreaIndex];
    bucketDataPointIndexes.push(previousBucketsSize + maxAreaIndex);

    previousBucketsSize += bucket.length;
  }

  bucketDataPointIndexes.push(previousBucketsSize);

  return bucketDataPointIndexes;
}

// Largest triangle three buckets data downsampling algorithm implementation
export const createLTTB = <P>(config: DownsamplingFunctionConfig<P>): DownsamplingFunction<P, [number]> => {
  const normalize = createNormalize(config.x, config.y);

  return (data: P[], desiredLength: number): P[] => {
    if (desiredLength < 0) {
      throw new Error(`Supplied negative desiredLength parameter to LTTB: ${desiredLength}`);
    }

    const { length } = data;
    if (length <= 1 || length <= desiredLength) return data;

    const normalizedData: NormalizedDataPoint[] = normalize(data);
    const buckets: NormalizedDataPoint[][] = splitIntoBuckets(normalizedData, desiredLength);
    const bucketDataPointIndexes: number[] = LTTBIndexesForBuckets(buckets);

    return bucketDataPointIndexes.map((index) => data[index]);
  };
};

export const LTTB = createLTTB(createLegacyDataPointConfig());
