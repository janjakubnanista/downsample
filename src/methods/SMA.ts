import { DownsamplingFunction, SmoothingFunctionConfig } from '../types';
import { createLegacyDataPointConfig, getPointValueExtractor } from '../utils';

export const SMANumeric = (data: number[], windowSize: number, slide = 1): number[] => {
  const output: number[] = [];
  let sum = 0;

  for (let i = 0; i < windowSize; i++) {
    sum += data[i];
  }

  for (let i = windowSize; i <= data.length; i++) {
    if ((i - windowSize) % slide === 0) {
      output.push(sum / windowSize);
    }

    sum += data[i] - data[i - windowSize];
  }

  return output;
};

/**
 * Simple Moving Average (SMA)
 *
 * @param data {Number[]}
 * @param windowSize {Number}
 * @param slide {Number}
 */
export const createSMA = <T>(
  config: SmoothingFunctionConfig<T>,
): DownsamplingFunction<T, [number, number | undefined] | [number]> => {
  const timeExtractor = getPointValueExtractor(config.x);
  const valueExtractor = getPointValueExtractor(config.y);
  const pointFactory = config.toPoint;

  return (values, windowSize, slide = 1): T[] => {
    if (values.length === 0) return [];
    // if (windowSize === 1 && slide === 1) return values.slice();

    const data: number[] = values.map(valueExtractor);
    const times: number[] = values.map(timeExtractor);
    const output: T[] = [];
    let sum = 0;

    for (let i = 0; i < windowSize; i++) {
      sum += data[i];
    }

    for (let i = windowSize; i <= data.length; i++) {
      if ((i - windowSize) % slide === 0) {
        output.push(pointFactory((times[i - windowSize] + times[i - 1]) / 2, sum / windowSize, i - windowSize));
      }

      sum += data[i] - data[i - windowSize];
    }

    return output;
  };
};

export const SMA = createSMA(createLegacyDataPointConfig());
