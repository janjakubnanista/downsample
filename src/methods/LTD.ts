import { DownsamplingFunction, DownsamplingFunctionConfig, Indexable, NormalizedDataPoint, Value } from '../types';
import { LTTBIndexesForBuckets } from './LTTB';
import { createLegacyDataPointConfig, createNormalize, iterableBasedOn, splitIntoBuckets } from '../utils';

export const mergeBucketAt = (buckets: NormalizedDataPoint[][], index: number): NormalizedDataPoint[][] => {
  const bucketA: NormalizedDataPoint[] = buckets[index];
  const bucketB: NormalizedDataPoint[] = buckets[index + 1];
  if (!bucketA || !bucketB) {
    throw new Error(`Bucket index out of range for merging: ${index} (allowed indexes are 0 - ${buckets.length - 2}`);
  }

  const mergedBucket: NormalizedDataPoint[] = [...bucketA, ...bucketB];

  const newBuckets = buckets.slice();
  newBuckets.splice(index, 2, mergedBucket);

  return newBuckets;
};

export const splitBucketAt = (buckets: NormalizedDataPoint[][], index: number): NormalizedDataPoint[][] => {
  const bucket: NormalizedDataPoint[] = buckets[index];
  if (!bucket) {
    throw new Error(`Bucket index out of range for splitting: ${index} (allowed indexes are 0 - ${buckets.length - 1}`);
  }

  const bucketSize: number = bucket.length;
  if (bucketSize < 2) {
    return buckets;
  }

  const bucketALength = Math.ceil(bucketSize / 2);
  const bucketA: NormalizedDataPoint[] = bucket.slice(0, bucketALength);
  const bucketB: NormalizedDataPoint[] = bucket.slice(bucketALength);

  const newBuckets = buckets.slice();
  newBuckets.splice(index, 1, bucketA, bucketB);

  return newBuckets;
};

export const calculateLinearRegressionCoefficients = (data: NormalizedDataPoint[]): [Value, Value] => {
  const N: number = data.length;

  let averageX: Value = 0;
  let averageY: Value = 0;
  for (let i = 0; i < N; i++) {
    averageX += data[i][0];
    averageY += data[i][1];
  }

  averageX /= N;
  averageY /= N;

  let aNumerator = 0;
  let aDenominator = 0;
  for (let i = 0; i < N; i++) {
    const [x, y]: NormalizedDataPoint = data[i];
    aNumerator += (x - averageX) * (y - averageY);
    aDenominator += (x - averageX) * (x - averageX);
  }

  const a: number = aNumerator / aDenominator;
  const b: number = averageY - a * averageX;

  return [a, b];
};

export const calculateSSEForBucket = (dataPoints: NormalizedDataPoint[]): number => {
  const [a, b] = calculateLinearRegressionCoefficients(dataPoints);

  let sumStandardErrorsSquared = 0;
  for (let i = 0; i < dataPoints.length; i++) {
    const dataPoint: NormalizedDataPoint = dataPoints[i];
    const standardError: number = dataPoint[1] - (a * dataPoint[0] + b);

    sumStandardErrorsSquared += standardError * standardError;
  }

  return sumStandardErrorsSquared;
};

export const calculateSSEForBuckets = (buckets: NormalizedDataPoint[][]): number[] => {
  // We skip the first and last buckets since they only contain one data point
  const sse: number[] = [0];

  for (let i = 1; i < buckets.length - 1; i++) {
    const previousBucket: NormalizedDataPoint[] = buckets[i - 1];
    const currentBucket: NormalizedDataPoint[] = buckets[i];
    const nextBucket: NormalizedDataPoint[] = buckets[i + 1];
    const bucketWithAdjacentPoints: NormalizedDataPoint[] = [
      previousBucket[previousBucket.length - 1],
      ...currentBucket,
      nextBucket[0],
    ];

    sse.push(calculateSSEForBucket(bucketWithAdjacentPoints));
  }

  sse.push(0);

  return sse;
};

export const findLowestSSEAdjacentBucketsIndex = (sse: number[], ignoreIndex: number): number | undefined => {
  let minSSESum: number = Number.MAX_VALUE;
  let minSSEIndex: number | undefined = undefined;
  for (let i = 1; i < sse.length - 2; i++) {
    if (i === ignoreIndex || i + 1 === ignoreIndex) {
      continue;
    }

    if (sse[i] + sse[i + 1] < minSSESum) {
      minSSESum = sse[i] + sse[i + 1];
      minSSEIndex = i;
    }
  }

  return minSSEIndex;
};

export const findHighestSSEBucketIndex = (buckets: NormalizedDataPoint[][], sse: number[]): number | undefined => {
  let maxSSE = 0;
  let maxSSEIndex: number | undefined = undefined;
  for (let i = 1; i < sse.length - 1; i++) {
    if (buckets[i].length > 1 && sse[i] > maxSSE) {
      maxSSE = sse[i];
      maxSSEIndex = i;
    }
  }

  return maxSSEIndex;
};

// Largest triangle three buckets data downsampling algorithm implementation
export const createLTD = <P>(config: DownsamplingFunctionConfig<P>): DownsamplingFunction<P, [number]> => {
  const normalize = createNormalize(config.x, config.y);

  return <Input extends Indexable<P> = Indexable<P>>(data: Input, desiredLength: number): Input => {
    if (desiredLength < 0) {
      throw new Error(`Supplied negative desiredLength parameter to LTD: ${desiredLength}`);
    }

    const { length } = data;
    if (length <= 2 || length <= desiredLength) {
      return data;
    }

    // Now we are sure that:
    //
    // - length is [2, Infinity)
    // - threshold is (length, Inifnity)
    const normalizedData: NormalizedDataPoint[] = normalize(data);

    // Require: data . The original data
    // Require: threshold . Number of data points to be returned
    // 1: Split the data into equal number of buckets as the threshold but have the first
    // bucket only containing the first data point and the last bucket containing only
    // the last data point . First and last buckets are then excluded in the bucket
    // resizing
    // 2: Calculate the SSE for the buckets accordingly . With one point in adjacent
    // buckets overlapping
    // 3: while halting condition is not met do . For example, using formula 4.2
    // 4: Find the bucket F with the highest SSE
    // 5: Find the pair of adjacent buckets A and B with the lowest SSE sum . The
    // pair should not contain F
    // 6: Split bucket F into roughly two equal buckets . If bucket F contains an odd
    // number of points then one bucket will contain one more point than the other
    // 7: Merge the buckets A and B
    // 8: Calculate the SSE of the newly split up and merged buckets
    // 9: end while.
    // 10: Use the Largest-Triangle-Three-Buckets algorithm on the resulting bucket configuration
    // to select one point per buckets

    let buckets: NormalizedDataPoint[][] = splitIntoBuckets(normalizedData, desiredLength);
    const numIterations = (length * 10) / desiredLength;
    for (let iteration = 0; iteration < numIterations; iteration++) {
      // 2: Calculate the SSE for the buckets accordingly . With one point in adjacent
      // buckets overlapping
      const sseForBuckets: number[] = calculateSSEForBuckets(buckets);

      // 4: Find the bucket F with the highest SSE
      const highestSSEBucketIndex: number | undefined = findHighestSSEBucketIndex(buckets, sseForBuckets);
      if (highestSSEBucketIndex === undefined) {
        break;
      }

      // 5: Find the pair of adjacent buckets A and B with the lowest SSE sum . The
      // pair should not contain F
      const lowestSSEAdajacentBucketIndex: number | undefined = findLowestSSEAdjacentBucketsIndex(
        sseForBuckets,
        highestSSEBucketIndex,
      );
      if (lowestSSEAdajacentBucketIndex === undefined) {
        break;
      }

      // 6: Split bucket F into roughly two equal buckets . If bucket F contains an odd
      // number of points then one bucket will contain one more point than the other
      buckets = splitBucketAt(buckets, highestSSEBucketIndex);

      // 7: Merge the buckets A and B
      // If the lowest SSE index was after the highest index in the original
      // unsplit array then we need to move it by one up since now the array has one more element
      // before this index
      buckets = mergeBucketAt(
        buckets,
        lowestSSEAdajacentBucketIndex > highestSSEBucketIndex
          ? lowestSSEAdajacentBucketIndex + 1
          : lowestSSEAdajacentBucketIndex,
      );
    }

    const dataPointIndexes: number[] = LTTBIndexesForBuckets(buckets);
    const outputLength = dataPointIndexes.length;
    const output: Input = iterableBasedOn(data, outputLength);
    for (let i = 0; i < outputLength; i++) {
      output[i] = data[dataPointIndexes[i]];
    }

    return output;
  };
};

export const LTD = createLTD(createLegacyDataPointConfig());
