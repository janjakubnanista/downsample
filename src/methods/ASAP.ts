import { DownsamplingFunction, Indexable, SmoothingFunctionConfig } from '../types';
import { SMANumeric, createSMA } from './SMA';
import { calculateMean, calculateSTD, createLegacyDataPointConfig, getPointValueExtractor, mapToArray } from '../utils';
import { fft, inverseFFT } from '../fft';

const calculateDiffs = (values: number[]): number[] => {
  const length = values.length - 1;
  if (length < 1) return [];

  const diffs = new Array(length);
  for (let i = 0; i < length; i++) {
    diffs[i] = values[i + 1] - values[i];
  }

  return diffs;
};

const calculateRoughness = (values: number[]): number => calculateSTD(calculateDiffs(values));

const calculateKurtosis = (values: number[]): number => {
  const length = values.length;
  const mean = calculateMean(values);

  let u4 = 0;
  let variance = 0;
  let diff: number;
  let diffSqr: number;
  for (let i = 0; i < length; i++) {
    diff = values[i] - mean;
    diffSqr = diff * diff;

    u4 += diffSqr * diffSqr;
    variance += diffSqr;
  }
  return (length * u4) / (variance * variance);
};

const findWindowSize = (
  head: number,
  tail: number,
  data: number[],
  minRoughness: number,
  originalKurt: number,
  windowSize: number,
): number => {
  while (head <= tail) {
    const w = Math.round((head + tail) / 2);
    const smoothed = SMANumeric(data, w, 1);
    const kurtosis = calculateKurtosis(smoothed);
    if (kurtosis >= originalKurt) {
      /* Search second half if feasible */
      const roughness = calculateRoughness(smoothed);
      if (roughness < minRoughness) {
        windowSize = w;
        minRoughness = roughness;
      }

      head = w + 1;
    } else {
      /* Search first half */
      tail = w - 1;
    }
  }

  return windowSize;
};

interface Autocorrelation {
  readonly peaks: number[];
  readonly correlations: number[];
  readonly maxCorrelation: number;
}

const calculatePeaks = (correlations: number[], threshold = 0.2): [number[], number] => {
  const { length } = correlations;
  if (length <= 1) return [[], 0];

  let maxCorrelation = 0;
  const peaks: number[] = [];

  if (correlations.length > 1) {
    let positive = correlations[1] > correlations[0];
    let max = 1;
    for (let i = 2; i < correlations.length; i += 1) {
      if (!positive && correlations[i] > correlations[i - 1]) {
        max = i;
        positive = !positive;
      } else if (positive && correlations[i] > correlations[max]) {
        max = i;
      } else if (positive && correlations[i] < correlations[i - 1]) {
        if (max > 1 && correlations[max] > threshold) {
          peaks.push(max);
          if (correlations[max] > maxCorrelation) {
            maxCorrelation = correlations[max];
          }
        }
        positive = !positive;
      }
    }
  }

  /* If there is no autocorrelation peak within the MAX_WINDOW boundary try windows from the largest to the smallest */
  if (peaks.length <= 1) {
    for (let i = 2; i < length; i += 1) {
      peaks.push(i);
    }
  }

  return [peaks, maxCorrelation];
};

const calculateAutocorrelation = (values: number[], maxLag: number): Autocorrelation => {
  const { length } = values;
  const mean = calculateMean(values);

  /* Padding to the closest power of 2 */
  const len = Math.pow(2, Math.trunc(Math.log2(length)) + 1);
  const fftreal = new Array(len).fill(0);
  const fftimg = new Array(len).fill(0);

  for (let i = 0; i < length; i += 1) {
    fftreal[i] = values[i] - mean;
  }

  /* F_R(f) = FFT(X) */
  fft(fftreal, fftimg);

  /* S(f) = F_R(f)F_R*(f) */
  for (let i = 0; i < fftreal.length; i += 1) {
    fftreal[i] = Math.pow(fftreal[i], 2) + Math.pow(fftimg[i], 2);
    fftimg[i] = 0;
  }

  /*  R(t) = IFFT(S(f)) */
  inverseFFT(fftreal, fftimg);

  // Calculate correlations
  const correlations: number[] = [];
  for (let i = 1; i < maxLag; i++) {
    correlations[i] = fftreal[i] / fftreal[0];
  }

  const [peaks, maxCorrelation] = calculatePeaks(correlations);

  return { correlations, peaks, maxCorrelation };
};

export const createASAP = <P>(config: SmoothingFunctionConfig<P>): DownsamplingFunction<P, [number]> => {
  const valueExtractor = getPointValueExtractor(config.y);
  const SMA = createSMA<P>(config);

  return function ASAP<Input extends Indexable<P> = Indexable<P>>(values: Input, resolution: number): Input {
    if (values.length === 0) return values;
    if (resolution <= 0) {
      throw new Error(`Supplied non-positive resolution parameter to ASAP: ${resolution}`);
    }

    // If the resolution is at least twice as big as the number of data points
    // then the values get downsampled to desired resolution first by SMA
    if (values.length >= 2 * resolution) {
      const scale = Math.trunc(values.length / resolution);

      return ASAP<Input>(SMA(values, scale, scale) as Input, resolution);
    }

    // First turn data points into a sequence of values
    const data: number[] = mapToArray(values, valueExtractor);

    const { correlations, peaks, maxCorrelation } = calculateAutocorrelation(data, Math.round(data.length / 10));
    const originalKurtosis = calculateKurtosis(data);
    let minRoughness = calculateRoughness(data);
    let windowSize = 1;
    let lb = 1;
    let largestFeasible = -1;
    let tail = data.length / 10;
    for (let i = peaks.length - 1; i >= 0; i -= 1) {
      const w = peaks[i];
      if (w < lb || w === 1) {
        break;
      } else if (Math.sqrt(1 - correlations[w]) * windowSize > Math.sqrt(1 - correlations[windowSize]) * w) {
        continue;
      }

      const smoothed = SMANumeric(data, w, 1);
      const kurtosis = calculateKurtosis(smoothed);
      const roughness = calculateRoughness(smoothed);
      if (kurtosis >= originalKurtosis) {
        if (roughness < minRoughness) {
          minRoughness = roughness;
          windowSize = w;
        }

        lb = Math.round(Math.max(w * Math.sqrt((maxCorrelation - 1) / (correlations[w] - 1)), lb));
        if (largestFeasible < 0) {
          largestFeasible = i;
        }
      }
    }

    if (largestFeasible > 0) {
      if (largestFeasible < peaks.length - 2) {
        tail = peaks[largestFeasible + 1];
      }
      lb = Math.max(lb, peaks[largestFeasible] + 1);
    }

    windowSize = findWindowSize(lb, tail, data, minRoughness, originalKurtosis, windowSize);

    return SMA(values, windowSize, 1) as Input;
  };
};

export const ASAP = createASAP(createLegacyDataPointConfig());
