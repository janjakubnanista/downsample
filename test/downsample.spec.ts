import 'jest';

import { ASAP, LTD, LTOB, LTTB, TupleDataPoint } from 'downsample';
import { ASAP as ASAP2 } from 'downsample/methods/ASAP';
import { LTD as LTD2 } from 'downsample/methods/LTD';
import { LTOB as LTOB2 } from 'downsample/methods/LTOB';
import { LTTB as LTTB2 } from 'downsample/methods/LTTB';
import downsample from 'downsample';

const NUMERIC_DATA = [1, 2, 3, 1, 5, 10, 15, 20];
const DATA_POINTS: TupleDataPoint[] = NUMERIC_DATA.map((value, index) => [index, value]);

describe('imports from root', () => {
  it('should have ASAP function', () => {
    expect(ASAP).toBeInstanceOf(Function);
    expect(ASAP(NUMERIC_DATA, 2)).toMatchSnapshot();
  });

  it('should have LTD function', () => {
    expect(LTD).toBeInstanceOf(Function);
    expect(LTD(DATA_POINTS, 0)).toMatchSnapshot();
  });

  it('should have LTOB function', () => {
    expect(LTOB).toBeInstanceOf(Function);
    expect(LTOB(DATA_POINTS, 0)).toMatchSnapshot();
  });

  it('should have LTTB function', () => {
    expect(LTTB).toBeInstanceOf(Function);
    expect(LTTB(DATA_POINTS, 0)).toMatchSnapshot();
  });
});

describe('default export', () => {
  it('should have ASAP function', () => {
    expect(downsample.ASAP).toBeInstanceOf(Function);
    expect(downsample.ASAP(NUMERIC_DATA, 2)).toMatchSnapshot();
  });

  it('should have LTD function', () => {
    expect(downsample.LTD).toBeInstanceOf(Function);
    expect(downsample.LTD(DATA_POINTS, 0)).toMatchSnapshot();
  });

  it('should have LTOB function', () => {
    expect(downsample.LTOB).toBeInstanceOf(Function);
    expect(downsample.LTOB(DATA_POINTS, 0)).toMatchSnapshot();
  });

  it('should have LTTB function', () => {
    expect(downsample.LTTB).toBeInstanceOf(Function);
    expect(downsample.LTTB(DATA_POINTS, 0)).toMatchSnapshot();
  });
});

describe('direct imports', () => {
  it('should have ASAP function', () => {
    expect(ASAP2).toBeInstanceOf(Function);
    expect(ASAP2([1, 10, 1, 10, 5], 2)).toMatchSnapshot();
  });

  it('should have LTD function', () => {
    expect(LTD2).toBeInstanceOf(Function);
    expect(LTD2(DATA_POINTS, 0)).toMatchSnapshot();
  });

  it('should have LTOB function', () => {
    expect(LTOB2).toBeInstanceOf(Function);
    expect(LTOB2(DATA_POINTS, 0)).toMatchSnapshot();
  });

  it('should have LTTB function', () => {
    expect(LTTB2).toBeInstanceOf(Function);
    expect(LTTB2(DATA_POINTS, 0)).toMatchSnapshot();
  });
});
