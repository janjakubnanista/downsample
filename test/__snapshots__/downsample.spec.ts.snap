// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default export should have ASAP function 1`] = `
Array [
  1.75,
  12.5,
]
`;

exports[`default export should have LTD function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    1,
    2,
  ],
]
`;

exports[`default export should have LTOB function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    7,
    20,
  ],
]
`;

exports[`default export should have LTTB function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    1,
    2,
  ],
]
`;

exports[`direct imports should have ASAP function 1`] = `
Array [
  5.5,
  5.5,
]
`;

exports[`direct imports should have LTD function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    1,
    2,
  ],
]
`;

exports[`direct imports should have LTOB function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    7,
    20,
  ],
]
`;

exports[`direct imports should have LTTB function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    1,
    2,
  ],
]
`;

exports[`imports from root should have ASAP function 1`] = `
Array [
  1.75,
  12.5,
]
`;

exports[`imports from root should have LTD function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    1,
    2,
  ],
]
`;

exports[`imports from root should have LTOB function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    7,
    20,
  ],
]
`;

exports[`imports from root should have LTTB function 1`] = `
Array [
  Array [
    0,
    1,
  ],
  Array [
    1,
    2,
  ],
]
`;
